### PR TITLE
refactor max_context_length

### LINKED ISSUE #4289: NoPE implementation and `max_seq_len`
## Problem

Currently, `GQAttention` has no way to support NoPE, which is a popular technique used in models such as `muse_glimmer`.

`muse_glimmer` implements NoPE by introducing a `use_rope: bool` option on the attention layer that can be used to disable RoPE application. While functional, this approach still requires the RoPE cache to exist, unnecessarily occupying memory.

At the config level, I think the cleanest way to support this would be to allow `rope` to be `None`, similarly to the existing norm configurations:

```python
class GQAttention(BaseAttention):

    @dataclass(kw_only=True, slots=True)
    class Config(BaseAttention.Config):
        n_heads: int
        dim: int
        ...
        qk_norm: RMSNorm.Config | None = None
        ...
        rope: RoPE.Config | None = None
```

However, this has some ramifications for the current design:

1. `Decoder` depends on the first attention layer to determine `max_context_length` (which is currently only used in tests).
2. `Decoder` walks all attention layers to set `rope.max_context_length`, which is required to build a correctly sized cache.

Allowing `rope` to be `None` would clash with both of these assumptions. We can work around them, but doing so raises a couple of design questions:

1. What should the maximum context length be when there is no RoPE? I think the correct answer is effectively `float("inf")`.
2. What happens if RoPE is used in the third attention layer rather than the first one?

This brings me to the broader point: I think the current `max_context_length` design is somewhat brittle.

## Proposed Solution

I think Titan's config system is great, probably the best I've seen. However, I'm not sure whether the intention is for the config system to catch configuration errors eagerly, or whether it should remain simple and let invalid configurations fail at runtime.

### Solution1: Crash at runtime

In this case, I would much prefer making the user responsible for setting `rope.max_context_length` appropriately for the trainer's context length.

If `rope.max_context_length` is too small, the model will simply fail at runtime as soon as that RoPE layer is executed. At most, we could add a runtime assertion ensuring that the RoPE context length is equal to the one in the batch (so that the cache is sufficient nor wasted).

This approach is simple and avoids adding configuration-time logic for what is ultimately.

### Solution2: Crash at config time

If we do want to validate this at config time, we need to decide where the rule

```text
trainer.context_length == rope.context_length
```

should be enforced.

Some possible options:

* **Trainer**: The trainer could walk the configuration tree, find all RoPE configurations, and enforce the constraint.
* **RoPE**: RoPE currently has no direct way to know about its parent or the trainer. We could modify the config system so that child configs have access to their parents, allowing RoPE to participate in this validation.
* **Decoder**: `Decoder` requires `max_context_length` to be set. Further, it walks its children to ensure that their `max_context_length` is set properly. 

Personally, I think failing at runtime is the better option here. It avoids introducing potentially complex configuration-time rules, while providing very little additional benefit: if the configuration is invalid, the model will fail anyway as soon as the affected RoPE layer is executed.

More generally, I'd prefer keeping the configuration system relatively simple and only enforcing constraints at config time when there is a strong reason to do so.

## Feedback

What do you think about this direction?

I'd be happy to contribute an implementation for supporting optional RoPE / NoPE, but I'd love to get some feedback on the preferred approach to handling the context-length constraint before starting.

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
### What

The model's context length is now decided when the spec is built, instead of being inferred from layer 0's RoPE config and retro-fitted onto RoPE caches later.

- `ModelSpec` gains a required `max_context_length`. `model_registry(flavor, seq_len=...)` sizes the RoPE caches up front; each flavor entry carries its own maximum, and asking for more raises `ValueError`.
- `Decoder.Config.update_from_config` no longer resizes RoPE caches or validates the training length, and `Decoder.Config.max_context_length` is gone.
- `seq_len` defaults to `None` in every config registry, meaning "the flavor's max context length".
- Callers updated to match: RL examples, graph_trainer, torchft, and the transformers modeling backend.
- Every CI entry point pins the length it used before.

### Why

PR to address https://github.com/pytorch/torchtitan/issues/4289

The context length is a property of the spec you build and is known before the trainer starts, so it should be stated there. Deriving it from one attention layer's RoPE config and mutating caches during `update_from_config` made it implicit, forced inference-only callers (the RL generator) to opt out of the sync, and left two places that could disagree about how long a sequence the model supports.

### CODE REVIEW

File: torchtitan/config/manager.py
Comment (tianyu-l) [inline torchtitan/config/manager.py:1]: let's not hack this file -- fyi we are moving away from CLI configuring https://github.com/pytorch/torchtitan/pull/4114

File: torchtitan/models/deepseek_v3/__init__.py [hunk 1]
```diff
@@ -619,17 +623,23 @@ def _671b(
 
 def model_registry(
     flavor: str,
+    *,
+    seq_len: int | None = None,
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/__init__.py:631]: I think the change in this PR makes it friendly for pretraining / midtraining, where it's legit to modify default max context length. This is what I liked about this PR's direction.

However, it smells dangerous for SFT when users could accidentally specify an input longer than checkpoint's max context length.

Is your point that SFT users should call `model_registry` without specifying this field? It gives no guard, basically.

File: torchtitan/experiments/rl/actors/trainer.py [hunk 2]
```diff
@@ -273,6 +273,12 @@ def _build_model(
         model_spec.model.update_from_config(config=config)
 
         # Check if the requested context exceeds the model context length.
+        # TODO: model_spec.model.max_context_length is the RoPE cache size, which
+        # core config_registry functions now build from their seq_len parameter
+        # rather than from the architecture's full context. Once a model registry
+        # in torchtitan/models is used here, this reads the training sequence
+        # length, not the model's generation capacity. Give ModelSpec a dedicated
+        # capacity field and read that instead.
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py]: Does it make sense if we still make `model.max_context_length` a cached property, by traversing through all `RoPE.max_context_length`? I'm considering deprecate `ModelSpec` and put all fields in `Model` / `Decoder` / their configs.

Or is there model that has absolutely no RoPE at all? In that case, what'd be the context length limit?

File: torchtitan/config/manager.py
Comment (francesco-bertolotti) [reply torchtitan/config/manager.py:1]: Thank you for letting me know. I think it is a good choice. We are not using the CLI either.

The way I like to use titan is directly with a trainer config:

```python
Trainer.Config(...).train()
```

It ends being a pretty large config file but I can easily swap, modify, and version any change.

File: torchtitan/models/deepseek_v3/__init__.py (hunk 1 above)
Comment (francesco-bertolotti) [reply torchtitan/models/deepseek_v3/__init__.py:631]: > However, it smells dangerous for SFT when users could accidentally specify an input longer than checkpoint's max context length.

You mean that users could do SFT with a context length larger then the one seen in pretrain? It would be odd but I do not see a reason to prevent it.

> Is your point that SFT users should call model_registry without specifying this field? It gives no guard, basically.

Actually, I think users should be entirely responsible of setting the `seq_len`. I think a default could be convenient. But if you think that this could be error prone, I can also remove it.

File: torchtitan/experiments/rl/actors/trainer.py (hunk 2 above)
Comment (francesco-bertolotti) [reply torchtitan/experiments/rl/actors/trainer.py]: In my opinion, the `max_context_length` is something that the users need to provide. Trying to infer what that could be is going to be error prone and architecture dependent.

File: torchtitan/config/manager.py
Comment (tianyu-l) [reply torchtitan/config/manager.py:1]: cc @fegin on feedback

File: torchtitan/models/deepseek_v3/__init__.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/__init__.py:631]: > You mean that users could do SFT with a context length larger then the one seen in pretrain? It would be odd but I do not see a reason to prevent it.

I'm always a bit worried that users would, by accident, specify over-long inputs to a trained model with intrinsic limitation on context length. Say model X has 10k context length capability, but user specifies 1M context input for SFT because their data is long.

Another way to put it, when model X is released with claimed max context length, how should a training framework like torchtitan respect this field? This PR is doing "set it as default in `model_registry` and let user freely override", which I feel may also be fine, because that's the more general thing to do.

I think the tension is really between
- torchtitan for pretraining (+post training) workflow
- torchtitan for post-training only workflow

also cc @fegin on the new `model_registry` arg

File: torchtitan/config/manager.py
Comment (fegin) [reply torchtitan/config/manager.py:1]: @francesco-bertolotti 

We are discussing how to do the transformation (e.g., enabling different CP backend, enabling different TP backend). Would like to hear users' opinion as well: https://github.com/pytorch/torchtitan/pull/4322#discussion_r3858313933

File: torchtitan/config/manager.py [hunk 3]
```diff
@@ -15,6 +16,10 @@
 
 from torchtitan.tools.logging import logger
 
+# Spellings tyro accepts for TrainingConfig.max_context_length. Kept in argv
```
Comment (tianyu-l) [inline torchtitan/config/manager.py]: let's remove hacks in this file

File: torchtitan/experiments/rl/actors/trainer.py
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py]: Let's remove this code, and also https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/decoder.py#L217-L241

File: torchtitan/models/llama3/__init__.py
Comment (tianyu-l) [inline torchtitan/models/llama3/__init__.py:384]: let's provide a per-flavor default max_context_length in this dictionary, and when calling `model_registry` if user given seq_len is longer than it, we raise

File: torchtitan/experiments/rl/actors/generator.py [hunk 4]
```diff
@@ -908,6 +908,12 @@ def __init__(
             # Enables RequestOutput.metrics, so generator metrics can be returned
             disable_log_stats=False,
         )
+        # TODO: model_spec.model.max_context_length is the RoPE cache size, which
+        # core config_registry functions now build from their seq_len parameter
+        # rather than from the architecture's full context. Once a model registry
+        # in torchtitan/models is used here, this reads the training sequence
+        # length, not the model's generation capacity. Give ModelSpec a dedicated
+        # capacity field and read that instead.
         engine_kwargs["max_model_len"] = model_spec.model.max_context_length
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py]: we should remove `model_spec.model.max_context_length`, right?

Comment (tianyu-l) [reply torchtitan/experiments/rl/actors/generator.py]: I saw that vllm also has "max_position_embeddings" required at https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/models/vllm_registry.py#L150

I think that should be kept -- it seems to be getting max context length from our rope field to set cache, which sounds right, except for that we will see models in the future without rope so probably needs to be optional

===============================================

Regarding this `max_model_len`, here are something meaningful after some codebase search
- general config https://github.com/vllm-project/vllm/blob/99013d77d332a2d21d7214b57fa495f2bad2b448/vllm/config/scheduler.py#L29
- For KV cache pre-allocation https://github.com/vllm-project/vllm/blob/99013d77d332a2d21d7214b57fa495f2bad2b448/docs/features/context_extension.md?plain=1#L69
- For maximum generation length https://github.com/vllm-project/vllm/blob/99013d77d332a2d21d7214b57fa495f2bad2b448/vllm/v1/engine/__init__.py#L54

It seems we should keep it the same as what our user sets for trainer (if you couldn't train longer than this, you shouldn't generate longer).

================================================

@francesco-bertolotti could you add a new field `ModelSpec.max_context_length` together with (`name`, `flavor`), which we set during `model_registry()` call and modify the code here to be
```suggestion
        engine_kwargs["max_model_len"] = model_spec.max_context_length
```

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/llama3/__init__.py
Comment (francesco-bertolotti) [reply torchtitan/models/llama3/__init__.py:384]: I would like to confirm if you mean something like this:

```python
llama3_configs = {
    "debugmodel": (_debugmodel, 2048),
    "1B": (_1b, 131072),
    "3B": (_3b, 131072),
    "8B": (_8b, 131072),
    "70B": (_70b, 131072),
    "405B": (_405b, 131072),
}


def model_registry(
    flavor: str,
    *,
    seq_len: int = 131072,
    attn_backend: str = "flex",
    tp_gemm_backend: TpGemmBackend = "default",
    converters: list[ModelConfigConverter.Config] | None = None,
) -> ModelSpec:
    # seq_len is forwarded only when the caller sets it, so each flavor
    # builder keeps its architecture's full context as the default.
    get_config, max_seq_len = llama3_configs[flavor]

    if seq_len > max_seq_len:
        logger.warning(
            f"Requested seq_len={seq_len} exceeds max context length for {flavor}."
        )
        max_seq_len = seq_len

    config = get_config(
        attn_backend=attn_backend,
        tp_gemm_backend=tp_gemm_backend,
        seq_len=max_seq_len
    )
```

That would be, when using `model_registry`:
- we choose a "max" `seq_len` value for rope_cache.
  - if users provide a smaller value we keep the default one. Not strictly a problem, the rope_cache will be bigger to accommodate any request. But why allocating more then neccessary?
  - if users provide a bigger value we accomodate the user.

Why not something like this:
```python
llama3_configs = {
    "debugmodel": _debugmodel,
    "1B": _1b,
    "3B": _3b,
    "8B": _8b,
    "70B": _70b,
    "405B": _405b,
}


def model_registry(
    flavor: str,
    *,
    seq_len: int = 8192,
    attn_backend: str = "flex",
    tp_gemm_backend: TpGemmBackend = "default",
    converters: list[ModelConfigConverter.Config] | None = None,
) -> ModelSpec:
    config = llama3_configs[flavor](
        attn_backend=attn_backend,
        tp_gemm_backend=tp_gemm_backend,
        **({} if seq_len is None else {"seq_len": seq_len})
    )
```
We would do exactly what the user say. If he does not say anything, we use our per-flavor default provided in the keyword argument of the respective `_1b`, ..., `_405b`

Comment (francesco-bertolotti) [reply torchtitan/models/llama3/__init__.py:384]: Third option:
```python
llama3_configs = {
    "debugmodel": (_debugmodel, 2048),
    "1B": (_1b, 131072),
    "3B": (_3b, 131072),
    "8B": (_8b, 131072),
    "70B": (_70b, 131072),
    "405B": (_405b, 131072),
}

def model_registry(
    flavor: str,
    *,
    seq_len: int | None,
    attn_backend: str = "flex",
    tp_gemm_backend: TpGemmBackend = "default",
    converters: list[ModelConfigConverter.Config] | None = None,
) -> ModelSpec:
    get_config, max_seq_len = llama3_configs[flavor]
    config = get_config(
        attn_backend=attn_backend,
        tp_gemm_backend=tp_gemm_backend,
        seq_len = seq_len or max_seq_len
    )
```

This would also allow to pass `seq_len` in the returned `ModelSpec` pretty easily. I think you meant something like this

Comment (tianyu-l) [reply torchtitan/models/llama3/__init__.py:384]: I meant third, yes. But I want the `seq_len > max_seq_len` check -- in particular, I want it to raise, not warning. My rationale is that
- when you use this particular model, it's capability is already intrinsically limited.
- but when you are doing pretraining of your own model, you don't need to have this raise / check.

**Comment by francesco-bertolotti** [comment]
`max_context_length` is now decided once, at `model_registry()` time, and carried on `ModelSpec.max_context_length`.

- Each `*_configs` entry becomes `(builder, max_context_length)`: the flavor's intrinsic context limit. Flavor builders take a required `seq_len` instead of hardcoding the RoPE length.
- `model_registry(flavor, seq_len=None)` uses the flavor cap when `seq_len` is omitted, and raises when the requested `seq_len` exceeds it.
- Removed the runtime RoPE resize + validation from `Decoder.Config.update_from_config`, the `Decoder.Config.max_context_length` property, and the duplicated check in the RL  `PolicyTrainer`.
- vLLM's `max_model_len` now reads `model_spec.max_context_length` instead of walking into the model config's RoPE.

File: torchtitan/models/deepseek_v3/config_registry.py [hunk 5]
```diff
@@ -206,9 +211,10 @@ def deepseek_v3_16b_minimal_async_ep() -> Trainer.Config:
     return config
 
 
-def deepseek_v3_671b() -> Trainer.Config:
+def deepseek_v3_671b(seq_len: int = 4096) -> Trainer.Config:
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/config_registry.py]: Given that we already have default in `model_registry`, let's make the default to be `None` for `config_registry.py` files.

**Comment by francesco-bertolotti** [comment]
seq_len now defaults to None in every config registry. If it is not specified, the model registry resolves it to the flavor's max context length, which is much larger than what these configs used before. 

That would have changed what CI runs, so I pinned seq_len at every CI entry point to the value it resolved to before this branch

File: torchtitan/experiments/transformers_modeling_backend/__init__.py [hunk 6]
```diff
@@ -160,11 +160,16 @@ class TitanMoeModelConfig(TitanModelConfig):
 }
 
 
-def model_registry(flavor: str) -> ModelSpec:
+_MAX_CONTEXT_LEN = 131072
```
Comment (tianyu-l) [inline torchtitan/experiments/transformers_modeling_backend/__init__.py]: how is this determined

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/experiments/transformers_modeling_backend/__init__.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/experiments/transformers_modeling_backend/__init__.py]: according to https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/transformers_modeling_backend/model.py#L462

Let's make `seq_len: int` required for `model_registry` for this particular modeling backend, and remove `_MAX_CONTEXT_LEN`.

Comment (francesco-bertolotti) [reply torchtitan/experiments/transformers_modeling_backend/__init__.py]: `seq_len` is now a required keyword arg on that model_registry and `_MAX_CONTEXT_LEN` is gone.
